File: ZLForm/Classes/ZLFormSectionDescriptor.m
//
//  ZLFormSectionDescriptor.m
//  ZLForm
//
//  Created by admin on 2026/5/15.
//

#import "ZLFormSectionDescriptor.h"
#import "ZLFormDescriptor.h"
#import "ZLFormRowDescriptor.h"
NSString * const ZLFormRowsKey = @"formRows";

@interface ZLFormSectionDescriptor ()
@property (nonatomic, strong,readwrite) NSMutableArray<ZLFormRowDescriptor *> *formRows;
@end
@implementation ZLFormSectionDescriptor
-(instancetype)init
{
    if (self = [super init]) {
        [self addObserver:self
               forKeyPath:ZLFormRowsKey
                  options:(NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld) context:0];
    }
    
    return self;
}
- (NSMutableArray *)formRows {
    if (!_formRows) {
        _formRows = [NSMutableArray array];
    }
    return _formRows;
}
-(instancetype)initWithTag:(NSString *)tag  {
    self = [self init];
    if (self) {
        self.tag = tag;
    }
    return self;
}

-(void)addFormRow:(nonnull ZLFormRowDescriptor *)formRow{
    if (![self.formRows containsObject:formRow]){
        formRow.sectionDescriptor = self;
        [self insertObject:formRow inFormRowsAtIndex:self.formRows.count];
    }
}
-(void)addFormRow:(ZLFormRowDescriptor *)formRow afterRow:(ZLFormRowDescriptor *)afterRow {
   NSInteger index = [self.formRows indexOfObject:afterRow];
    if (index != NSNotFound) {
        formRow.sectionDescriptor = self;
        [self insertObject:formRow inFormRowsAtIndex:index + 1];
    }
}

-(void)addFormRow:(ZLFormRowDescriptor *)formRow beforeRow:(ZLFormRowDescriptor *)beforeRow {
    NSInteger index = [self.formRows indexOfObject:beforeRow];
    if (index != NSNotFound) {
        formRow.sectionDescriptor = self;
        [self insertObject:formRow inFormRowsAtIndex:index];
    }
}

-(void)removeFormRow:(ZLFormRowDescriptor *)formRow {
    NSInteger index = [self.formRows indexOfObject:formRow];
    if (index != NSNotFound) {
        [self removeObjectFromFormRowsAtIndex:index];
    }
    
}

-(void)removeFormRowWithTag:(NSString *)tag {
    ZLFormRowDescriptor *rowDescriptor = [self formRowWithTag:tag];
    if (rowDescriptor) {
        [self removeFormRow:rowDescriptor];
    }
}
- (ZLFormRowDescriptor *)formRowWithTag:(NSString *)tag {
    ZLFormRowDescriptor *rowDescriptor = nil;
    for (ZLFormRowDescriptor *row in self.formRows) {
        if ([row.tag isEqualToString:tag]) {
            rowDescriptor = row;
            break;
        }
    }
    return rowDescriptor;
}

#pragma mark - KVC

-(NSUInteger)countOfFormRows
{
    return self.formRows.count;
}

- (id)objectInFormRowsAtIndex:(NSUInteger)index
{
    return [self.formRows objectAtIndex:index];
}

- (NSArray *)formRowsAtIndexes:(NSIndexSet *)indexes
{
    return [self.formRows objectsAtIndexes:indexes];
}

- (void)insertObject:(ZLFormRowDescriptor *)formRow inFormRowsAtIndex:(NSUInteger)index
{
    formRow.sectionDescriptor = self;
    [self.formRows insertObject:formRow atIndex:index];
}

- (void)removeObjectFromFormRowsAtIndex:(NSUInteger)index
{
    [self.formRows removeObjectAtIndex:index];
}


#pragma mark - KVO

- (void)observeValueForKeyPath:(nullable NSString *)keyPath
                      ofObject:(nullable id)object
                        change:(nullable NSDictionary<NSKeyValueChangeKey, id> *)change
                       context:(nullable void *)context
{
     if ([keyPath isEqualToString:ZLFormRowsKey]) {
        if ([self.formDescriptor.formSections containsObject:self]) {
            if ([[change objectForKey:NSKeyValueChangeKindKey] isEqualToNumber:@(NSKeyValueChangeInsertion)]) {
                NSIndexSet *indexSet = [change objectForKey:NSKeyValueChangeIndexesKey];
                ZLFormRowDescriptor *formRow = [((ZLFormSectionDescriptor *)object).formRows objectAtIndex:indexSet.firstIndex];
                NSUInteger sectionIndex = [self.formDescriptor.formSections indexOfObject:object];
                [self formRowHasBeenAdded:formRow
                                                      atIndexPath:[NSIndexPath indexPathForRow:indexSet.firstIndex inSection:sectionIndex]];
               
            }
            else if ([[change objectForKey:NSKeyValueChangeKindKey] isEqualToNumber:@(NSKeyValueChangeRemoval)]) {
                NSIndexSet *indexSet = [change objectForKey:NSKeyValueChangeIndexesKey];
                ZLFormRowDescriptor *removedRow = [[change objectForKey:NSKeyValueChangeOldKey] objectAtIndex:0];
                NSUInteger sectionIndex = [self.formDescriptor.formSections indexOfObject:object];
                [self formRowHasBeenRemoved:removedRow
                                                        atIndexPath:[NSIndexPath indexPathForRow:indexSet.firstIndex inSection:sectionIndex]];
            }
        }
    }
}
-(void)dealloc
{
    [self removeObserver:self forKeyPath:ZLFormRowsKey];
    [self.formRows removeAllObjects];
    self.formRows = nil;
}


-(void)formRowHasBeenAdded:(ZLFormRowDescriptor *)formRow atIndexPath:(NSIndexPath *)indexPath
{
    
    UITableView *tableView = self.formDescriptor.tableView;
    [tableView beginUpdates];
    [tableView insertRowsAtIndexPaths:@[indexPath] withRowAnimation:formRow.insertAnimation];
    [tableView endUpdates];
    if ([self.formDescriptor.delegate respondsToSelector:@selector(formDescriptor:formRowHasBeenAdded:atIndexPath:)]) {
        [self.formDescriptor.delegate formDescriptor:self.formDescriptor formRowHasBeenAdded:formRow
                                              atIndexPath:[NSIndexPath indexPathForRow:indexPath.row inSection:indexPath.section]];
    }
}

-(void)formRowHasBeenRemoved:(ZLFormRowDescriptor *)formRow atIndexPath:(NSIndexPath *)indexPath
{
  
    UITableView *tableView = self.formDescriptor.tableView;
    [tableView beginUpdates];
    [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:formRow.deleteAnimation];
    [tableView endUpdates];
    if ([self.formDescriptor.delegate respondsToSelector:@selector(formDescriptor:formRowHasBeenRemoved:atIndexPath:)]) {
        [self.formDescriptor.delegate formDescriptor:self.formDescriptor formRowHasBeenRemoved:formRow
                                                atIndexPath:[NSIndexPath indexPathForRow:indexPath.row inSection:indexPath.section]];
    }
}

@end

File: ZLForm/Classes/ZLFormSectionDescriptor.swift
import UIKit
import DifferenceKit
public typealias ZLFormSectionHeaderFooterViewBlock = (_ sectionDescriptor: ZLFormSectionDescriptor) -> UIView?

@objcMembers
public class ZLFormSectionDescriptor: NSObject,Differentiable {
    
    public var headerHeight: CGFloat = 0
    public var headerViewBlock: ZLFormSectionHeaderFooterViewBlock?
    
    public var footerHeight: CGFloat = 0
    public var footerViewBlock: ZLFormSectionHeaderFooterViewBlock?
    
    public var sectionBackgroundView: UIView?
    
    public var sectionBackgroundInsets: UIEdgeInsets = .zero
    
    public weak var formDescriptor: ZLFormDescriptor?
    
    public private(set) var formRows: [ZLFormRowDescriptor] = []
    
    public var hidden: Bool = false
    
    /// section 展示出来的时候赋值，默认为 -1，表示未赋值。ZLFormDescriptor 会在展示 section 的时候赋值为当前 section 在 formSections 中的 index。
    public var section: Int = -1

    public var tag: String = ""
    
    public var title: String?
    
    @objc public dynamic var value: Any?


    // MARK: - Init
    
    private override init() {
        super.init()
    }
    
    public init(tag: String) {
        super.init()
        self.tag = tag
    }
    
    // MARK: - Row Management
    
    public func addFormRow(_ formRow: ZLFormRowDescriptor) {
        guard !formRows.contains(formRow) else { return }
        formRow.sectionDescriptor = self
        let index = formRows.count
        formRows.append(formRow)
        notifyRowAdded(formRow, at: index)
    }
    
    public func addFormRow(_ formRow: ZLFormRowDescriptor, afterRow: ZLFormRowDescriptor) {
        guard let index = formRows.firstIndex(of: afterRow) else { return }
        formRow.sectionDescriptor = self
        let insertIndex = index + 1
        formRows.insert(formRow, at: insertIndex)
        notifyRowAdded(formRow, at: insertIndex)
    }
    
    public func addFormRow(_ formRow: ZLFormRowDescriptor, beforeRow: ZLFormRowDescriptor) {
        guard let index = formRows.firstIndex(of: beforeRow) else { return }
        formRow.sectionDescriptor = self
        formRows.insert(formRow, at: index)
        notifyRowAdded(formRow, at: index)
    }
    
    public func removeFormRow(_ formRow: ZLFormRowDescriptor) {
        guard let index = formRows.firstIndex(of: formRow) else { return }
        formRows.remove(at: index)
        notifyRowRemoved(formRow, at: index)
    }
    
    public func removeFormRow(tag: String) {
        guard let row = formRow(tag: tag) else { return }
        removeFormRow(row)
    }
    
    public func formRow(tag: String) -> ZLFormRowDescriptor? {
        return formRows.first { $0.tag == tag }
    }
    
    // MARK: - Notifications
    
    private func notifyRowAdded(_ formRow: ZLFormRowDescriptor, at index: Int) {
        guard let descriptor = formDescriptor else { return }
        descriptor.reloadDiff()
        if let sectionIndex = descriptor.formSections.firstIndex(where: { $0.model === self }) {
            let indexPath = IndexPath(row: index, section: sectionIndex)
            descriptor.delegate?.formDescriptor?(descriptor, formRowHasBeenAdded: formRow, at: indexPath)
        }
    }
    
    private func notifyRowRemoved(_ formRow: ZLFormRowDescriptor, at index: Int) {
        guard let descriptor = formDescriptor else { return }
        descriptor.reloadDiff()
        if let sectionIndex = descriptor.formSections.firstIndex(where: { $0.model === self }) {
            let indexPath = IndexPath(row: index, section: sectionIndex)
            descriptor.delegate?.formDescriptor?(descriptor, formRowHasBeenRemoved: formRow, at: indexPath)
        }
    }
    
    // MARK: - Sorting
    
    /// Sorts formRows by tag. Called by ZLFormDescriptor when sortByTag is true.
    public func sortRowsByTagIfNeeded() {
        formRows.sort { $0.tag < $1.tag }
    }
    
    // MARK: - Differentiable
    private let uuid = UUID().uuidString

    public var differenceIdentifier: String {
        return tag.isEmpty ? uuid : tag
    }
    public var contentEqualHandler: ((ZLFormSectionDescriptor, ZLFormSectionDescriptor) -> Bool)?

    public func isContentEqual(to source: ZLFormSectionDescriptor) -> Bool {
        if let handler = contentEqualHandler ?? source.contentEqualHandler {
            return handler(self, source)
        }
        return tag == source.tag &&
               title == source.title &&
               (value as? NSObject)?.isEqual(source.value) == true

    }
}
